codice/ddf
# [2.13.x] DDF-4132 Use internal URLs for CAS redirects

#### What does this PR do?
Reverts the changes from PR #3481 and reverts the CAS client version to 3.2.2 due to a bug in 3.4.1.
#### Select relevant component teams: 
@codice/security 
#### Ask 2 committers to review/merge the PR and tag them here.
@brendan-hofmann @vinamartin @stustison 
#### How should this be tested?
Testing this PR is going to be difficult because I'm not aware of unzip-and-run proxy testing infrastructure that meets the needs here. Ideally, it should be tested with a proxy configured to:
1. Rewrite URLs in HTTP redirects
2. NOT send original request host to proxied server. For Apache, turn `ProxyPreserveHost` off, and in Nginx, don't set the HTTP Host header (i.e. don't use `proxy_set_header Host ...`)
3. Proxy DDF behind a path prefix (e.g. https://proxy:443/ddf/admin -> https://ddf:8993/admin)
You can try to use the changes in LinkMJB/nginx-reverse-proxy#3 which will allow you to do (3). But that won't support proxying both CAS and DDF. Also, as-is it won't do (1), and even if you configure rewriting (i.e. using the proxy_redirect config), running the proxy in a Docker container causes issues because the port in the rewritten URL is not the port that the user uses to hit the proxy (the proxy doesn't know about Docker and it's forwarded ports)

I'd suggest running an Nginx instance outside of Docker with a configuration like the following:
```
        location /ddf/ {
            proxy_pass         https://ddf:8993/;
            proxy_set_header   X-Real-IP $remote_addr;
            proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header   X-Forwarded-Host $server_name;
        }  
        location /cas/ {
            proxy_pass         https://cas:8443/cas/;
            proxy_set_header   X-Real-IP $remote_addr;
            proxy_set_header   X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header   X-Forwarded-Host $server_name;
        } 
```
Then you just need to:
* Set up CAS. This _might_ help: https://github.com/codice/cas-distribution
* Install DDF and modify the `system.properties` to specify the correct internal/external hosts
* Configure trust for Nginx and DDF
* Configure trust for CAS and Nginx

With all that working, verify that you can login via CAS. Also verify that you can login with or without using trailing slashes. 

#### Any background context you want to provide?
We experienced issues with the CAS client using external URLs when constructing redirects. In particular, those changes relied on the proxy sending the external hostname in the HTTP Host header, but doing that broke Jetty trailing slash redirects. This PR resulted from a bit of discussion about what DDF should actually be responsible for in a proxy setup. In the case of the CAS client, all the redirect rewriting can easily be performed by a proxy, so it makes sense to avoid the added complexity of computing redirects ourselves and the headache of making sure it works for both proxy and non-proxy setups.
#### What are the relevant tickets?
[DDF-4132](https://codice.atlassian.net/browse/DDF-4132)
#### Screenshots

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
